src/runtimeHelpers.ts

```diff
@@ -253,6 +298,141 @@ export function unPartitionResults<T>(
+ * ],
+ *
+ *
+ * IMPORTANT NOTE: The contract must have a one-to-one correspondence between the input propertyBatchKey and the output propertyBatchKey.
+ * i.e. if we have property: 'name' in the request, the response must have 'name' in it, and no extra data assciated with it.
+ *
+ * Example
+ * Request args:
+ * [
+ *      { bar_id: 2, properties: ['name'], include_extra_info: true },
+ *      { bar_id: 1, properties: ['rating'], include_extra_info: false },
+ *      { bar_id: 2, properties: ['rating'], include_extra_info: true },
+ * ]
+ *
+ * ```js
+ * unPartitionResultsByBatchKeyPartition(
+ *   newKey = 'bar_id',
+ *   propertyBatchKey =  'properties',
+ *   batchKeyPartition = [ [2, 2], [1] ],
+ *   propertyBatchKeyPartion = [ [['name'], ['rating']], [['rating']] ],
+ *   requestGroups = [ [0, 2], [1] ],
+ *   resultGroups = [
+ *      [ { bar_id: 2, name: 'Burger King', rating: 3 } ],
+ *      [ { bar_id: 1, name: 'In N Out', rating: 4 } ]
+ *   ],
+ * )
+ * ```
+ *
+ * Returns:
+ * ```
+ * [
+ *   { bar_id: 2, name: 'Burger King' }，
+ *   { bar_id: 1, rating: 4 }，
+ *   { bar_id: 2, rating: 3 }，
+ * ]
+ */
+export function unPartitionResultsByBatchKeyPartition<T extends Record<string, any>>(
+    newKey: string,
+    propertyBatchKey: string,
+    batchKeyPartition: ReadonlyArray<ReadonlyArray<any>>,
+    propertyBatchKeyPartion: ReadonlyArray<ReadonlyArray<any>>,
+    /** Should be a nested array of IDs, as generated by partitionItems */
+    requestGroups: ReadonlyArray<ReadonlyArray<number>>,
+    /** The results back from the service, in the same shape as groups */
+    resultGroups: ReadonlyArray<ReadonlyArray<T | CaughtResourceError>>,
+): ReadonlyArray<T | Error> {
+    /**
+     * e.g. with our inputs, produce:
+     * ```js
+     * [
+     *   [
+     *      { order: 0, result: { bar_id: 2, name: 'Burger King' },
+     *      { order: 2, result: { bar_id: 2, rating: 3 } },
+     *   ],
+     *   [
+     *      { order: 1, result: { bar_id: 1, rating: 4 } },
+     *   ]
+     * ]
+     * ```
+     */
+    const zippedGroups: ReadonlyArray<ReadonlyArray<{ order: number; result: T | Error }>> = requestGroups.map(
+        (ids, i) => {
+            return ids.map((id, j) => {
+                let result = null;
+                for (const resultElement of Object.values(resultGroups)[i]) {
+                    // There's error in the result we should return
+                    if (resultElement instanceof CaughtResourceError) {
+                        result = resultElement;
+                        break;
+                    }
+
+                    if (resultElement[newKey] === batchKeyPartition[i][j]) {
+                        result = Object.assign(
+                            {},
+                            ...[newKey, ...propertyBatchKeyPartion[i][j]].map((key) => ({
+                                [key]: resultElement[key],
+                            })),
+                        );
```

Comment:
1. the logic here is quite dense with multiple levels of spreading....can we split this up at all?
2. `Object.assign` is ok but unnecessary- you can use straight up object literal spread: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals

Reply:
I used `_pick` instead.